### PR TITLE
fix(base) reduce the amount of timers used

### PR DESCRIPTION
**Pull request by kikito**
The current implementation creates one recurring timer per balancer.

This is ok in many circumstances, but can exhaust nginx's allocated
resources.

This PR changes that by a single timer which updates all balancers
sequentially. When there's a need for updating a host, *that* part
gets delegated to an ephemeral (non-recurring) timer.

This way only 1 recurring timer gets created, and ephemeral
timers are created only when hosts need to be updated, and then
released.

### CODE REVIEW

**Comment by Tieske** [comment]
why create the ephermeral timer? why not execute in-line?

More structural comment; we should not be registering balancers, but hosts. The hosts should be put in a binary tree, by ttl, and once per second we should check the tree. With large sets of balancers that will be far more efficient, since we only need to check a few hosts that expired, instead of looping over all of them (assume 1600 balancers, each with 5 hosts, that is 8000 entries we would iterate over every second)

**Comment by kikito** [comment]
> why create the ephermeral timer? why not execute in-line?

My assumption is that `host:queryDns` is not immediate - that it takes some time. If it takes 1 cent of a second to execute, then that means that if more than 100 hosts need to be updated, then doing them "in-line" would mean that one execution of the main timer would not have finished when the next one starts.

By delegating each call to `host:queryDns` to an ephemeral timer, the main timer is pure non-blocking Lua. It can go over I don't know how many hosts per second. Hundreds of thousands?. So this should avoid the "step over itself" problem.

There can still be some "wasteful timer" if `host:queryDns` takes longer than 1 second for a given host. Then the first main timer will create 1 ephemeral timer to queryDns, the second one will create another, etc, until the first one finishes. But they will eventually clean up.

> More structural comment; we should not be registering balancers, but hosts. The hosts should be put in a binary tree, by ttl, and once per second we should check the tree. With large sets of balancers that will be far more efficient, since we only need to check a few hosts that expired, instead of looping over all of them (assume 1600 balancers, each with 5 hosts, that is 8000 entries we would iterate over every second)

My concern was fixing the issue with the client that we were currently having, not a general solution.

**Comment by Tieske** [comment]
> My assumption is that host:queryDns is not immediate - that it takes some time. If it takes 1 cent of a second to execute, then that means that if more than 100 hosts need to be updated, then doing them "in-line" would mean that one execution of the main timer would not have finished when the next one starts.

lua-resty-timer reschedules the next invocation after the work is done, so this would not happen. AT least that was my thinking, but that is only the case with the most recent version, the previous version would actually reschedule before starting the work.

Anyway, I think this should be closed in favour of a better solution

**Comment by kikito** [comment]
Ok, closing for now.

**Comment by kikito** [comment]
I still think that the "a single recurring thread which checks everything and delegates jobs to ephemeral threads" is the way to go here. Optimizing the checks with a tree data structure would certainly help, but if we make it fast enough, it might not be needed. "Marking" the hosts which are already being processed so no wasteful ephemeral threads are needed would also be important in order to make this work for everyone.

I also forgot to mention one main advantage of this approach vs a more "sequential" one: doing things on ephemeral timers allows for a parallel processing of the "queue", provided that there are workers available.

I would advice to try to not rely on a "glocal" var (`_entities`, on my example). I would very much rather have that "list of entities" living somewhere on the user's space. But this would need an extra entity, like a "connection". I don't know if such thing exists on this library. If not, that would be a breaking change.

File: src/resty/dns/balancer/base.lua [hunk 1]
```diff
@@ -1367,24 +1353,47 @@ _M.new = function(opts)
...
+  local function check_for_expired_records()
+    for balancer in pairs(_balancers) do
+      --check all hosts for expired records,
+      --including those with errors
+      --we update, so changes on the list while traversing can happen, keep track of that
+      for _, host in ipairs(balancer.hosts) do
+        -- only retry the errored ones
+        if ((host.lastQuery or EMPTY).expire or 0) < time() then
+          ngx_log(ngx_DEBUG, balancer.log_prefix, "executing requery for: ", host.hostname)
+          ngx.timer.at(0, delayed_host_querydns, host)
```
Comment (thibaultcha) [inline src/resty/dns/balancer/base.lua:1380]: Here we need to catch the error from the timer creation as well, if any.

File: src/resty/dns/balancer/base.lua [hunk 2]
```diff
@@ -160,6 +160,8 @@ local errors = setmetatable({
 })
 
 
+local _balancers = setmetatable({}, { __mode = 'v' })
```
Comment (thibaultcha) [inline src/resty/dns/balancer/base.lua:163]: Shouldn't the mode be `k`?

**Pull request closed without merging**